channels/offline-channel-js/src/OfflineBatchHandler.ts

```diff
@@ -0,0 +1,387 @@
+            };
+            
+
+            _self.storeBatch = (batch: IPayloadData, cb?: OfflineBatchStoreCallback, sync?: boolean) => {
+                
+                let provider = _provider;
+                if(!!sync) {
+                    provider = _unloadProvider;
+                }
+                return createAsyncPromise((resolve, reject) => {
+                    if (!provider) {
+                        reject(new Error(NoProviderErrMsg))
+                    }
+                    let evt = _getOfflineEvt(batch);
+                    return doAwaitResponse(provider.addEvent(evt.id, evt, _itemCtx), (response: AwaitResponse<IStorageTelemetryItem>) => {
+                        try {
+                            let evt = response.value || response.reason || [];
+                            let state = eBatchStoreStatus.Failure;
+                            if (!response.rejected) {
+                                state = eBatchStoreStatus.Success;
+                            }
+                            let res = {state: state, item: evt};
+                            cb && cb(res);
+                            resolve(res);
+                        } catch(e) {
+                            reject(e);
+                        }
+                    });
+                });
+
+            }
+
+            _self.sendNextBatch = (cb?: OfflineBatchCallback, sync?: boolean, xhrOverride?: IXHROverride, cnt: number = 1) => {
+                return createAsyncPromise((resolve, reject) => {
+                    let provider = _provider;
+                    try {
+                       
+                        if (_unloadProvider) {
+                            doAwaitResponse(_unloadProvider.getNextBatch(), (res) => {
+                                if (res && !res.rejected) {
+                                    let unloadEvts = res.value;
+                                    if (unloadEvts && unloadEvts.length) {
+                                        // always try to send ones from local storage
+                                        provider = _unloadProvider;
+                                    }
+    
+                                }
+                            });
+
+                        }
+                       
+                    } catch(e) {
+                        // eslint-disable-next-line no-empty
+                    }
+                    if (!provider) {
+                        reject(new Error(NoProviderErrMsg));
+                    }
+
+                    return doAwaitResponse(provider.getNextBatch(), (response: AwaitResponse<IStorageTelemetryItem[]>) => {
+                        try {
+                            if(!response || response.rejected) {
+                                let res = {state: eBatchSendStatus.Failure, data: response.reason};
+                                storeResolve(res);
+                                return;
```

Comment:
This should be calling either the resolve / reject from the outer promise -- otherwise it will never be fulfilled, so anyone waiting on the original to complete will wait forever.

Reply:
As below the `storeResolve` is calling the `resolve`